swift-sdk/Internal/InAppManager.swift

```diff
@@ -136,62 +133,140 @@ class InAppManager : NSObject, IterableInAppManagerProtocolInternal {
     }
     
     func set(read: Bool, forMessage message: IterableInAppMessage) {
-        updateQueue.sync {
-            let toUpdate = message
-            toUpdate.read = read
-            self.messagesMap.updateValue(toUpdate, forKey: message.messageId)
-            persister.persist(self.messagesMap.values)
+        updateMessage(message, read: true).onSuccess { _ in
+            self.callbackQueue.async {
+                self.notificationCenter.post(name: .iterableInboxChanged, object: self, userInfo: nil)
+            }
         }
-        self.callbackQueue.async {
-            self.notificationCenter.post(name: .iterableInboxChanged, object: self, userInfo: nil)
+    }
+    
+    func isOkToShowNow(message: IterableInAppMessage) -> Bool {
+        guard displayer.isShowingInApp() == false else {
+            ITBInfo("showing another")
+            return false
+        }
+        guard message.didProcessTrigger == false else {
+            ITBInfo("message with id: \(message.messageId) is already processed")
+            return false
+        }
+        guard getInAppShowingWaitTimeInterval() <= 0 else {
+            ITBInfo("can't display within retryInterval window")
+            return false
         }
+        
+        return true
     }
-
+    
     @objc private func onAppEnteredForeground(notification: Notification) {
         ITBInfo()
         let waitTime = InAppManager.getWaitTimeInterval(fromLastTime: lastSyncTime, currentTime: dateProvider.currentDate, gap: moveToForegroundSyncInterval)
         if waitTime <= 0 {
-            synchronize()
+            _ = scheduleSync()
         } else {
             ITBInfo("can't sync now, need to wait: \(waitTime)")
         }
     }
+
+    private func scheduleSync() -> Future<Bool, Error> {
+        let result = Promise<Bool, Error>()
+        self.syncQueue.async {
+            if let syncResult = self.syncResult {
+                if syncResult.isResolved() {
+                    self.syncResult = self.synchronize()
+                } else {
+                    self.syncResult = syncResult.flatMap { _ in self.synchronize() }
+                }
+            } else {
+                self.syncResult = self.synchronize()
+            }
+            self.syncResult?.onSuccess { _ in
+                result.resolve(with: true)
```

Comment:
This appears to only resolve the Future with a successful result, but not pass errors.